File: problems/123_best-time-to-buy-and-sell-stock-iii/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int buy1 = INT_MIN, sell1 = 0;
        int buy2 = INT_MIN, sell2 = 0;

        for (int p : prices)
        {
            buy1 = max(buy1, -p);         // best after 1st buy
            sell1 = max(sell1, buy1 + p); // best after 1st sell
            buy2 = max(buy2, sell1 - p);  // best after 2nd buy
            sell2 = max(sell2, buy2 + p); // best after 2nd sell
        }
        return sell2;
    }
    int maxProfit_2(vector<int> &prices)
    {
        // next[canBuy][txLeft], curr[canBuy][txLeft]
        int next[2][3] = {}; // initialized to 0
        int curr[2][3] = {};

        for (int i = (int)prices.size() - 1; i >= 0; --i)
        {
            for (int txLeft = 1; txLeft <= 2; ++txLeft)
            {
                // canBuy = 1
                curr[1][txLeft] = max(
                    -prices[i] + next[0][txLeft],
                    next[1][txLeft]);

                // canBuy = 0
                curr[0][txLeft] = max(
                    prices[i] + next[1][txLeft - 1],
                    next[0][txLeft]);
            }
            // move curr -> next
            for (int b = 0; b < 2; ++b)
                for (int t = 0; t < 3; ++t)
                    next[b][t] = curr[b][t];
        }
        return next[1][2];
    }
    int maxProfit_3(vector<int> &prices)
    {
        int n = (int)prices.size();

        // dp[i][canBuy][txLeft]
        vector<vector<vector<int>>> dp(n + 1,
                                       vector<vector<int>>(2, vector<int>(3, 0)));

        // dp[n][*][*] already 0; dp[*][*][0] already 0

        for (int i = n - 1; i >= 0; --i)
        {
            for (int txLeft = 1; txLeft <= 2; ++txLeft)
            {
                // canBuy = 1
                dp[i][1][txLeft] = max(
                    -prices[i] + dp[i + 1][0][txLeft], // buy
                    dp[i + 1][1][txLeft]               // skip
                );

                // canBuy = 0 (holding)
                dp[i][0][txLeft] = max(
                    prices[i] + dp[i + 1][1][txLeft - 1], // sell (consume 1 tx)
                    dp[i + 1][0][txLeft]                  // hold
                );
            }
        }

        return dp[0][1][2];
    }
    int maxProfit_4(vector<int> &prices)
    {
        int n = prices.size();
        vector<vector<vector<int>>> dp(n, vector<vector<int>>(2, vector<int>(3, INT_MIN)));
        return rec(prices, 0, 1, 2, dp);
    }
    int rec(vector<int> &prices, int idx, int canBuy, int txLeft,
            vector<vector<vector<int>>> &dp)
    {
        if (idx == (int)prices.size() || txLeft == 0)
            return 0;

        int &ans = dp[idx][canBuy][txLeft];
        if (ans != INT_MIN)
            return ans;

        if (canBuy)
        {
            int buy = -prices[idx] + rec(prices, idx + 1, 0, txLeft, dp);
            int skip = rec(prices, idx + 1, 1, txLeft, dp);
            ans = max(buy, skip);
        }
        else
        {
            int sell = prices[idx] + rec(prices, idx + 1, 1, txLeft - 1, dp);
            int hold = rec(prices, idx + 1, 0, txLeft, dp);
            ans = max(sell, hold);
        }
        return ans;
    }
};

int main()
{
    Solution solution;

    vector<pair<vector<int>, int>> samples = {
        {{3, 3, 5, 0, 0, 3, 1, 4}, 6},
        {{1, 2, 3, 4, 5}, 4},
        {{7, 6, 4, 3, 1}, 0},
    };

    for (size_t i = 0; i < samples.size(); ++i)
    {
        vector<int> prices = samples[i].first;
        int expected = samples[i].second;
        int result = solution.maxProfit(prices);
        cout << "Example " << (i + 1) << " result: " << result
             << ", expected: " << expected << '\n';
    }

    return 0;
}

File: problems/123_best-time-to-buy-and-sell-stock-iii/main.rs
struct Solution;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        // O(1) state machine solution (same as C++ `maxProfit`)
        let mut buy1 = i32::MIN;
        let mut sell1 = 0i32;
        let mut buy2 = i32::MIN;
        let mut sell2 = 0i32;

        for p in prices {
            buy1 = buy1.max(-p);           // best after 1st buy
            sell1 = sell1.max(buy1 + p);   // best after 1st sell
            buy2 = buy2.max(sell1 - p);    // best after 2nd buy
            sell2 = sell2.max(buy2 + p);   // best after 2nd sell
        }
        sell2
    }

    pub fn max_profit_2(prices: Vec<i32>) -> i32 {
        // Tabulated DP with rolling arrays (same as C++ `maxProfit_2`)
        let n = prices.len();
        let mut next = [[0i32; 3]; 2];
        let mut curr = [[0i32; 3]; 2];

        for i in (0..n).rev() {
            for tx_left in 1..=2 {
                // canBuy = 1
                curr[1][tx_left] = (-prices[i] + next[0][tx_left]).max(next[1][tx_left]);
                // canBuy = 0
                curr[0][tx_left] = (prices[i] + next[1][tx_left - 1]).max(next[0][tx_left]);
            }
            next = curr; // copy small fixed-size array
        }

        next[1][2]
    }

    pub fn max_profit_3(prices: Vec<i32>) -> i32 {
        // Full DP table (same as C++ `maxProfit_3`)
        let n = prices.len();
        let mut dp = vec![vec![vec![0i32; 3]; 2]; n + 1];

        for i in (0..n).rev() {
            for tx_left in 1..=2 {
                // canBuy = 1
                dp[i][1][tx_left] = (-prices[i] + dp[i + 1][0][tx_left]).max(dp[i + 1][1][tx_left]);
                // canBuy = 0
                dp[i][0][tx_left] = (prices[i] + dp[i + 1][1][tx_left - 1]).max(dp[i + 1][0][tx_left]);
            }
        }

        dp[0][1][2]
    }

    pub fn max_profit_4(prices: Vec<i32>) -> i32 {
        // Top-down recursion + memo (same as C++ `maxProfit_4` + `rec`)
        if prices.is_empty() {
            return 0;
        }
        let n = prices.len();
        let mut dp = vec![vec![vec![i32::MIN; 3]; 2]; n];
        Self::rec(&prices, 0usize, 1usize, 2usize, &mut dp)
    }

    fn rec(prices: &Vec<i32>, idx: usize, can_buy: usize, tx_left: usize, dp: &mut Vec<Vec<Vec<i32>>>) -> i32 {
        if idx == prices.len() || tx_left == 0 {
            return 0;
        }

        if dp[idx][can_buy][tx_left] != i32::MIN {
            return dp[idx][can_buy][tx_left];
        }

        let ans = if can_buy == 1 {
            let buy = -prices[idx] + Self::rec(prices, idx + 1, 0, tx_left, dp);
            let skip = Self::rec(prices, idx + 1, 1, tx_left, dp);
            buy.max(skip)
        } else {
            let sell = prices[idx] + Self::rec(prices, idx + 1, 1, tx_left - 1, dp);
            let hold = Self::rec(prices, idx + 1, 0, tx_left, dp);
            sell.max(hold)
        };

        dp[idx][can_buy][tx_left] = ans;
        ans
    }
}

fn main() {
    let samples: Vec<(Vec<i32>, i32)> = vec![
        (vec![3, 3, 5, 0, 0, 3, 1, 4], 6),
        (vec![1, 2, 3, 4, 5], 4),
        (vec![7, 6, 4, 3, 1], 0),
    ];

    for (idx, (prices, expected)) in samples.iter().enumerate() {
        let p = prices.clone();
        let r1 = Solution::max_profit(p.clone());
        let r2 = Solution::max_profit_2(p.clone());
        let r3 = Solution::max_profit_3(p.clone());
        let r4 = Solution::max_profit_4(p.clone());

        println!(
            "Example {} -> max_profit: {}, max_profit_2: {}, max_profit_3: {}, max_profit_4: {} (expected {})",
            idx + 1,
            r1,
            r2,
            r3,
            r4,
            expected
        );
    }
}
